Frameworks/Social.Xaml/ShareSheet.xaml.cpp

```diff
@@ -0,0 +1,111 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include "pch.h"
+#include <wrl/client.h>
+#include <wrl/implements.h>
+#include <ppltasks.h>
+#include "ShareSheet.xaml.h"
+#include "ShareSheetControl.h"
+#include "ErrorHandling.h"
+
+using namespace Social::Xaml;
+using namespace Microsoft::WRL;
+using namespace Windows::Foundation;
+using namespace Windows::UI::Xaml;
+using namespace Windows::UI::Xaml::Controls;
+using namespace Windows::UI::Xaml::Media;
+using namespace Windows::UI::Xaml::Media::Imaging;
+using namespace Windows::Web::Http;
+using namespace concurrency;
+
+ShareSheet::ShareSheet(Platform::String^ token) {
+    InitializeComponent();
+    _token = token;
+}
+
+void ShareSheet::ShareSheet_Post(ContentDialog^ sender, ContentDialogButtonClickEventArgs^ args) {
+    // Rudimentary implementation of posting logic
+    Platform::String^ urlStr = "https://graph.facebook.com/v2.8/me/feed"
+        "?message=" + Uri::EscapeComponent(Message->Text) +
+        "&privacy=" + Uri::EscapeComponent("{\"value\":\"ALL_FRIENDS\"}") +
+        "&access_token=" + Uri::EscapeComponent(_token);
+
+    if (_linkToShare != nullptr) {
+        urlStr += "&link=" + Uri::EscapeComponent(_linkToShare);
+    }
+
+    Uri^ url = ref new Uri(urlStr);
+    IHttpContent^ content = ref new HttpStringContent("");
+    HttpClient^ client = ref new HttpClient;
+
+    auto post = create_task(client->PostAsync(url, content));
+    post.then([] (HttpResponseMessage^ response) {
+        // Here's where we would notify our client of the result
+    });
+    
+}
+
+void ShareSheet::ShareSheet_Cancel(ContentDialog^ sender, ContentDialogButtonClickEventArgs^ args) {
+    // Notify caller of cancellation
+}
+
+void ShareSheet::addLink(Platform::String^ url) {
+    // What we should do here is to download a preview of the URL
+    // and render it into the Preview image. But that seems like work,
+    // so instead we'll just display the text of the URL.
+
+    TextBlock^ text = ref new TextBlock;
+    text->Text = url;
+
+    Preview->Children->Append(text);
+    _linkToShare = url;
+}
+
+void ShareSheet::addImage(ImageSource^ imageSource) {
+    Image^ image = ref new Image;
+    image->Source = imageSource;
+    Preview->Children->Append(image);
+    _imageToShare = image;
+}
+
+class ShareSheetImpl : public RuntimeClass<RuntimeClassFlags<ClassicCom>, ShareSheetControl> {
+    ShareSheet^ _shareSheet;
+
+public:
+    ShareSheetImpl(const wchar_t* token) {
+        _shareSheet = ref new ShareSheet(Platform::StringReference(token));
+    }
+
+    void addLink(const wchar_t* url) override {
+        _shareSheet->addLink(Platform::StringReference(url));
+    }
+
+    void addImage(IInspectable* image) override {
+        _shareSheet->addImage(reinterpret_cast<ImageSource^>(image));
```

Comment:
>reinterpret_cast [](start = 30, length = 16)

safe_cast<ImageSource^>(reinterpret_cast<Platform::Object^>(image)) might be safer?